### ISSUE TITLE
Eager produces n + 1 requests with belongs_to

### ISSUE THREAD

**Issue by stanislas-m** [gobuffalo/pop#46]
Using the following:

```go
type Room struct {
    ID       int          `db:"id"`
    // [...]
    Provider RoomProvider `belongs_to:"room_provider"`
}
```

```go
rooms := models.Rooms{}
q := tx.Eager("Provider")
if err := q.All(&rooms); err != nil {
    return err
}
```

The SQL produced is 1 query for the main request, then n requests (one for each row) to fetch providers. In the case of 1 to 1 relations, it should use `LEFT JOIN`, which is more efficient.

More reference about this issue: https://use-the-index-luke.com/sql/join/nested-loops-join-n1-problem

**Comment by gavbaa** [comment]
I think this would also be valuable for 1 to many relations, as the link describes.  Having to do N queries for relatively simple relations and small result set returns, the network time far outstrips the extra processing time.

While I can't speak for all active record implementations, some ORM frameworks definitely make this decision, including your referenced link and Django's ORM.  SQLAlchemy also can do 1-to-many like this.

**Comment by chenghung** [comment]
from my understanding, the active record rubygem implement eager load using below two queries
```
select * from rooms ....;
select * from room_providers where room_id IN (1,3,5, ....);
``` 
first, it fetch all rooms from db and collect distinct room ids 
then it fetch room providers using `IN (...)` query
finally it bind room provider orm object to corresponding room object.

it seems GORM implment eager load in the same way.
http://doc.gorm.io/crud.html#preloading-eager-loading

**Comment by mikkergimenez** [comment]
Has anyone started tackling this one yet?  I'd be game for it as it's currently impacting a project I'm working on.

**Comment by gavbaa** [comment]
I haven't had time to do something about this yet, unfortunately!  I'd be happy to test it out on my own project and offer input as you work through ideas.

**Comment by u007** [comment]
are you saying that the query only does 1 query?
i dont recommend this. this is not the way rails active record works.
once you do a left join, you will have problem with other select fields and where condition because it will start asking you to put in full table namespace to select and duplicate record may happen if there are multiple parent (which i know shouldn't be, but will happen)

**Comment by gavbaa** [comment]
It is the way the ActiveRecord implementation in the link works, as well as the way that several very prominent ActiveRecord implementations work.  Django's is probably the most well-known, and they just using table aliasing to build the query.  There's no issues with multiple parents since each inclusion of each table is aliased to a unique name (t1, t2, t3, etc.).  This behavior is far more useful for the average case than requiring users who are already intentionally asking for Eager behavior to have to still do their own 1-N query batches with IN clauses.  Most of the uses of Eager we see in our own application implementations are not 1-1, they're 1-N.

**Comment by u007** [comment]
@stanislas-m is referring to using only 1 query as in there will be no sub query with IN clauses.
am i right?

**Comment by gavbaa** [comment]
As far as I read, yes.  I meant the situation where query 1 gets the list of IDs you want to select, and a separate query 2 use IN clauses to get the related tables of records, that's what I wanted to see avoided for 1-N relations.

**Comment by mclark4386** [comment]
Thinking about it you are going to do loops over the results either way... I think going for the more end-user usable approach (run the original `All`, grab ids, do where on _other table_ with IN of the ids and then merge the results into our original objects) makes the most sense.

**Comment by gavbaa** [comment]
That strategy works very poorly performance-wise for returning result sets across nested tables, as you're now generating N queries worth of IN clauses.  To refer to the original link, my expectation would be:

```Criteria c = session.createCriteria(Employees.class);
c.add(Restrictions.ilike("lastName", 'Win%'));

c.setFetchMode("sales", FetchMode.JOIN);
c.setResultTransformer(Criteria.DISTINCT_ROOT_ENTITY);

List<Employees> result = c.list();
It generates the following query:

select this_.subsidiary_id as subsidiary1_0_1_
     , this_.employee_id as employee2_0_1_
       -- MORE this_ columns on employees
     , sales2_.sale_id as sale1_3_
       -- MORE sales2_ columns on sales
  from employees this_ 
  left outer join sales sales2_ 
          on this_.subsidiary_id=sales2_.subsidiary_id 
         and this_.employee_id=sales2_.employee_id 
 where lower(this_.last_name) like ?```

One query, where the related tables can have N results in them, and the construction of those results back to models happens as part of the Eager fetch.  This would make has_one and has_many operate just like belongs_to, and offers great benefits in not creating excessively-long SQL statements with giant IN clauses, or having to create N/batch_size queries to avoid excessively-long SQL statements.  If the user really wants to split things into IN clauses, that's something they can do independently, but they can't build up the mapping of the result set of a nested Eager query by themselves.

**Comment by u007** [comment]
oh no, arent we already using in clause for the eager part?

**Comment by gavbaa** [comment]
Yes, that's exactly what I was hoping this ticket would be fixing.

**Comment by mikkergimenez** [comment]
To summarize: the goal of this fix is to reduce the n+1 queries, by performing two queries, the first ALL query to get a list of records, and a second IN query to get a list of records which match the id's returned by the first query.

Currently, After an all query is performed, the Eager method iterates over the rows returned by all and performs an individual select = query for each of the rows.  

My approach was going to be to create a separate boolean method(say "Includes" since that matches the rails syntax, but something like "Accumulate" might be more accurate that tells eager to do the IN query rather than iterate over each row.  I was then going to modify/refactor the "eagerAssocations" method in finders.go to accommodate this difference.   

I could simply make this the default behavior, but I guess I wonder why Rails doesn't have this as the default behavior?

**Comment by mclark4386** [comment]
So I'm curious, do all dialects support sub-queries? Because if they do then instead of doing the id collect phase after the all and building a giant id array IN query for the second, we could basically have the first query be a sub-query for the second which could make things easier.

By way of the above example:
1st query: `select * from rooms` 
2nd query: `select * from providers where id in (select provider_id  from rooms)` (note that we could use the same original statement for the sub-query with the only change being in the projection, so any wheres, limits, etc would cross over as well)
The results of this can be loaded into a map by id to make faster associations in the joining loop.

I'm just thinking out load... ^^

**Comment by u007** [comment]
in rails it will collect all ids and make it a another query with In clauses (without left join).
and this is done by paginated result, meaning 1 eager query per page.

@mclark4386 sounds good, but im still thinking how it will effect complicated queries from original query

**Comment by mclark4386** [comment]
I don't _think_ anyone has started any code for this yet, so please mention any complicated queries you think of that my be an issue @u007 and others! Will make for good test cases for whoever ends up working on it!

**Comment by mikkergimenez** [comment]
I've got some code written for this, nothing pretty.  I can push my branch up if anyone wants to take a look.  I was able to rewrite the function so that it would perform the one query (I think it does in fact use the subquery that @mclark4386 mentioned above) but where I'm at now is that it's associating all the join records with the last record from the initial query.  

I think this is because the associations each include a pointer to the original record, and some of the magic behind the query automagically associates the join records with the record included in the association.  

It seems like there are two paths forward:

1: Create an intermediary that pulls the needed info out of the association, run the query independently (I _think_ this reference is what's linking the join query to the original all query) https://github.com/gobuffalo/pop/blob/master/finders.go#L208  and then stich the two back together. 

 The other would be to dig into the internals of the query and association, which I would imagine would be the better designed solution, but I may need some help with that spelunking.

**Comment by larrymjordan** [comment]
Guys I have worked on a proposal a couple of week ago,  I've decided to share my progress with all of you. Please take a look when you have the chance and share your thoughts. This issue has got a lot of attention lately and I really would like to contribute on something to solve this.
https://github.com/gobuffalo/pop/pull/146

**Comment by u007** [comment]
hi @mclark4386 i think its not possible to query based on parent query. we do have eager support for custom query. therefore the ideal way todo it will be alike rails, pick all current ids and put it as a where clause for subsequence query

**Comment by stanislas-m** [comment]
This should be fixed with #146. Thanks @larrymjordan!
